### ISSUE TITLE
Userset Rewrites with Intersections

### ISSUE THREAD

**Issue by jon-whit** [ory/keto#466]
The Zanzibar paper mentions that the userset expression language specification supports sub-expressions such as union, intersection, and exclusion.

One use case that I immediately have is to be able to express relation tuples over disjoint sets. In my model I have permission boundaries (analogous to [AWS IAM Permission Boundaries](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_boundaries.html)) and shared/inherited permissions. Permission boundaries specify the maximum level of permissions that a user can have, but it does not specify _what_ permissions the user has. Shared/inherited permissions define the permissions granted to the user either via ABAC policies or explicitly shared rights, but do not fully define the effective permissions the user has.

An access evaluation decision looks like this:

```
boundaries = {"projects.list", "projects.create"}
permissions = {"projects.list"}

 // The effective permissions is the set intersection of the two.
effectivePermissions = boundaries & permissions

effectivePermissions --> {"projects.list"}
```

I'd like to collaborate with you guys on coming up with the namespace configuration expression language that will allow us to express any arbitrary combo of sub-expressions involving unions, intersections, and exclusions (negations).

I've started with the work from [mishudark/zanzibar](https://github.com/mishudark/zanzibar). He's done a pretty good job at laying a foundation down for userset rewrites with unions, but it lacks support for intersection and exclusion. I also don't think that the yaml modeling that he's defined will expand very well once we allow abitrarily nested sub-expressions, so it's probably worth revisiting and going back to the drawing board.

**Comment by zepatrik** [comment]
What about a config similar to this:

![IMG_20210305_174831](https://user-images.githubusercontent.com/5354445/110146682-22cb7100-7ddb-11eb-9ad0-d0165486ba80.jpg)

This is just interesting from a content perspective, not syntax.

**Comment by jon-whit** [comment]
It's less around the configuration syntax and more around the evaluation of such rules.

I've been looking at [Google's Common Expression Language](https://opensource.google/projects/cel) and their go implementation [google/cel-go](https://github.com/google/cel-go). The question is how do we both define the syntax for rewrites and also evaluate arbitrary expressions composing those rewrites for Check and Expand evaluations.

My thinking around using CEL is that our userset rewrite configuration could just translate into a CEL expression, each operand of the expression can be evaluated using the Graph traversal, and then the operands can be fed into the CEL evaluation engine for an outcome.

**Comment by jon-whit** [comment]
From CEL's documentation:

> Ideal applications are ones in which the configuration is executed often and modified relatively infrequently.

This fits in with userset rewrite configurations as well. The expressions that result from the userset rewrites shouldn't change that frequently, but every Check or Expand RPC would have to evaluate such expressions. So CEL would potentially be a good candidate for such a use case.

**Comment by zepatrik** [comment]
CEL does look really promising, it might be useful elsewhere as well. Did you look at #468 already? I think it should be possible to implement it like that. Basically I modify the graph traversal algorithm to return on different conditions based on the rewrite rules.
But I will definitely dig deeper into CEL.

**Comment by jon-whit** [comment]
@zepatrik oh sorry I haven't take a look yet. Just throwing out ideas (sorry for the spam, haha).

**Comment by zepatrik** [comment]
No, CEL is definitely an awesome pointer, it seems even suitable as a config language from a first glance.

**Comment by aeneasr** [comment]
CEL looks really interesting!

**Comment by jon-whit** [comment]
@zepatrik I wanted to describe my use case a little better... The example above is non ideal for a few reasons that I outline below.

We have a data model where each user has an access level, and each access level has one or more permissions that are included in it.

```
type User struct {
    ID string
    AccessLevelID string
}

type AccessLevel struct {
    ID string
}

type AccessLevelPermissions struct {
    AccessLevelID string // joined on AccessLevel.ID
    IncludedPermissions []string
}
```

What I'd like to be able to do is write a userset rewrite that says "If the user has an access level that includes a certain (required) permission AND has any one of any number of shared permissions, then they are granted access". Additional, when an AccessLevel changes such that the permissions included in it change, that should minimally impact the number of relation tuples stored. If we denormalize the AccessLevelPermissions, then any time the AccessLevel changes every single relation tuple referencing it would have to change, and that would really suck.

Do you have any thoughts?

**Comment by zepatrik** [comment]
Actually implemented with #984
